tools/lkl/lib/virtio_net.c

```diff
@@ -66,32 +66,55 @@ static inline int is_tx_queue(struct virtio_dev *dev, struct virtio_queue *queue
        return &dev->queue[TX_QUEUE_IDX] == queue;
 }
 
-static int net_enqueue(struct virtio_dev *dev, struct virtio_req *req)
+/* Inside the LKL kernel, we wrapped this buffer (which corresponds to
+ * an IP packet) with a header to make sure it got routed to the TAP
+ * device. This header is meaningless to the oustide world, so we cut
+ * it off the top of the packet before sending it to the TAP
+ * device. */
+void strip_dev_buf_header(const struct lkl_dev_buf *in_buf,
+			struct lkl_dev_buf *out_buf,
+			struct lkl_virtio_net_hdr_v1 **out_hdr)
 {
-	struct lkl_virtio_net_hdr_v1 *header;
-	struct virtio_net_dev *net_dev;
-	int ret, len;
-	void *buf;
+	memset(out_buf, 0, sizeof(*out_buf));
 
-	header = req->buf[0].addr;
-	net_dev = netdev_of(dev);
-	len = req->buf[0].len - sizeof(*header);
+	/* Header starts at the top of the buffer */
+	*out_hdr = in_buf->addr;
+
+	out_buf->len = in_buf->len - sizeof(**out_hdr);
+	out_buf->addr = in_buf->addr + sizeof(**out_hdr);
+}
 
-	buf = &header[1];
+static int net_enqueue(struct virtio_dev *dev, struct virtio_req *req)
+{
+	struct lkl_virtio_net_hdr_v1 *header;
+	struct lkl_dev_buf out_buf;
+	int ret;
 
-	if (!len && req->buf_count > 1) {
-		buf = req->buf[1].addr;
-		len = req->buf[1].len;
+	struct virtio_net_dev *net_dev = netdev_of(dev);
+
+	/* Let's try the first buffer first */
+	strip_dev_buf_header(&req->buf[0], &out_buf, &header);
+
+	/* The first buffer was just a header. If this request has
+	 * multiple buffers, it's a TX packet and the info we care
+	 * about is in the second buffer (because an empty packet is
+	 * useless to the TAP device), so let's send that. If we don't
+	 * have multiple buffers, this must be an RX packet, and its
```

Comment:
For TX, it's LKL kernel sending out packets to the outside world, so the `out_buf` here contains actual data.  With GSO enabled inside LKL kernel, I believe they are jumbo frames.

For RX, the `out_buf` is a buffer to hold potential incoming packet data.  In the current implementation, the number of scatter-gather buffers is restricted to 1, but in the future we probably can have more complicated things for GRO.